File: euler/p23.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef uint64_t nb;
typedef uint8_t b;

int main() {
  nb somme = 0;
  nb n = 28124;

  // On calcule la somme des diviseurs de chaque nombre en O(n + des poussières)
  // Ainsi que les nombres abondants pour le même prix
  nb* divisors = calloc(n, sizeof(nb));
  b* abundant = calloc(n, sizeof(b));

  for (nb i = 1; i < n; i++) {
    if (divisors[i] > i) abundant[i] = 1;
    for (nb j = 2; i * j < n ; j++) {
      divisors[i * j] += i;
    }
  }

  // On calcule les nombres se décomposant comme deux nombres abondants
  // Ce qui donne une complexité de O(n^2)
  for (nb i = 1; i < n; i++) {
    somme += i;
    
    for (nb j = 12; j < i; j++) {
      if (abundant[j] == 1 && abundant[i - j] == 1) {
        somme -= i;
        break;
      }
    }
  }

  printf("Somme : %lu\n", somme);
}

File: euler/p23.ml

let somme n =
  let divisors = Array.make n 0 in
  let abundant = Array.make n 0 in
  (* On calcule les nombres abondants *)
  for i = 1 to n - 1 do
    if divisors.(i) > i then abundant.(i) <- 1;
    for j = 2 to (n - 1) / i do
      divisors.(i * j) <- divisors.(i * j) + i;
    done;
  done;
  (* On calcule la somme *)
  let somme = ref 0 in
  for i = 1 to n - 1 do
    let exists = ref false in
    for j = 12 to i do
      if abundant.(j) = 1 && abundant.(i - j) = 1
      then exists := true;
    done;
    if !exists = false then somme := !somme + i;
  done;
  !somme
